panoramapublic/src/org/labkey/panoramapublic/PanoramaPublicController.java

```diff
@@ -5214,6 +5434,241 @@ public void setDoi(String doi)
+                                            .addParameter("id", form.getId())
+                                            .addParameter("testAccount", Boolean.TRUE))
+                                    .build()
+                    ));
+        }
+
+        @Override
+        public void addNavTrail(NavTree root)
+        {
+            root.addChild("Post to Bluesky");
+        }
+    }
+
+    @RequiresPermission(AdminOperationsPermission.class)
+    public static class PostToBlueskyAction extends ConfirmAction<BlueskyForm>
+    {
+        private ExperimentAnnotations _expAnnot;
+        private BlueskyException _exception;
+        private String _blueskyAtUri;
+
+        @Override
+        public void validateCommand(BlueskyForm form, Errors errors)
+        {
+            _expAnnot = form.lookupExperiment();
+            if(_expAnnot == null)
+            {
+                errors.reject(ERROR_MSG, "Cannot find experiment with ID " + form.getId());
+                return;
+            }
+            ensureCorrectContainer(getContainer(), _expAnnot.getContainer(), getViewContext());
+            if(!_expAnnot.isJournalCopy())
+            {
+                errors.reject(ERROR_MSG, "Posting to Bluesky is only allowed on experiments in the Panorama Public project");
+                return;
+            }
+
+            if (_expAnnot.getShortUrl() == null)
+            {
+                errors.reject(ERROR_MSG, "Cannot create a post on Bluesky for an experiment that does not have a short URL. (Experiment id: " + _expAnnot.getId() + ").");
+                return;
+            }
+
+            JournalSubmission js = SubmissionManager.getSubmissionForJournalCopy(_expAnnot);
+            if (js == null)
+            {
+                errors.reject(ERROR_MSG, "Cannot find a submission request for copied experiment Id " + _expAnnot.getId());
+                return;
+            }
+            if (!js.isLatestExperimentCopy(_expAnnot.getId()))
+            {
+                errors.reject(ERROR_MSG, "Experiment id " + _expAnnot.getId() + " is not the last copied submission. "
+                        + getActionName(this.getClass()) + " is only allowed in the last copy of the submitted data");
+                return;
+            }
+
+            if (!_expAnnot.isPublic())
+            {
+                errors.reject(ERROR_MSG, "Cannot create a post on Bluesky for an experiment that is not public.");
+            }
+        }
+
+        @Override
+        public boolean handlePost(BlueskyForm form, BindException errors)
+        {
+            BlueskySettings settings = BlueskySettingsManager.getSettings();
+            try
+            {
+                // Post to Bluesky
+                _blueskyAtUri = BlueskyApiClient.getInstance().createPost(_expAnnot, settings, form.isTestAccount());
+            }
+            catch (BlueskyException e)
+            {
+                _exception = e;
+                LOG.error(e);
```

Comment:
I think this will just get you the `toString()`. Maybe that's enough here, but you might want the two-argument version which will get you the full stack trace